cumulus_library_opioid/vocab/static_builder.py

```diff
@@ -0,0 +1,131 @@
+import csv
+import dataclasses
+import pathlib
+
+import pandas
+
+from cumulus_library import base_table_builder, base_utils, study_manifest
+from cumulus_library.template_sql import base_templates
+
+@dataclasses.dataclass
+class TableConfig:
+    file_path: str
+    delimiter: str
+    table_name: str
+    headers: list[str]
+    dtypes: dict
+    parquet_types: list[str]
+    filtered_name: str | None = None
+    ignore_header: bool = False
+    map_cols: list[dict] | None = None
+
+class StaticBuilder(base_table_builder.BaseTableBuilder):
+    display_text = "Building static data tables..."
+    base_path = pathlib.Path(__file__).resolve().parent
+
+    tables = [
+        TableConfig(
+            base_path / './common/keywords/keywords.curated.tsv',
+            '\t',
+            'keywords',
+            ['STR'],
+            {"STR": 'str'},
+            ['STRING'],
+            filtered_name= base_path / './common/keywords/keywords.curated.filtered.tsv',
+        ),
+        TableConfig(
+            base_path / './common/expand_rules/expand_rules.tsv',
+            '\t',
+            'expand_rules',
+            # TODO: do we actually need to preserve 'rule'?
+            ["TTY1","RELA","TTY2","rule", ],
+            {"TTY1":'str',"RELA":'str',"TTY2":'str',"rule":'str'},
+            ['STR','STR','STR','STR','STR'],
+            ignore_header=True,
+            map_cols = [{'from':'rule','to':'include',"map_dict":{ 'yes': True, 'no':False }}]
+        )
+    ]
+
+    def filter_duplicated(self, path, delimiter, filtered_name):
+        target = path.with_suffix(".filtered.tsv")
+        #if target.exists():
+        #    return
+        with open(path) as file:
+            reader = csv.reader(file, delimiter=delimiter)
+            rows = []
+            for row in reader:
+                rows.append(row)
+        rows.sort(key=lambda x:len(x[1]))
+        unique_keywords =[]
+        while len(rows) !=0:
+            unique_keywords.append(rows[0][1])
+            keyword = rows[0][1]
+            rows.pop(0)
+            matching_indexes=[]
+            for index in range(0, len(rows)):
+                if keyword.casefold() in rows[index][1].casefold():
+                    matching_indexes.insert(0, index)
+            for index in matching_indexes:
+                rows.pop(index)
```

Comment:
I don't think this works - indexes will change after each call to `pop`.

But stepping back - something like this feels more readable:
```
keywords = {row[1] for row in reader}
unique_keywords = {k.casefold(): k for k in keywords}.values()
```

Reply:
Ah right you are.

I stand by my claim that this 30-odd lines of subtle index ordering would be best replaced with two lines.

Reply:
yeah you're not wrong there

Reply:
ok, so this is not actually that straightforward - let me add this as a comment to this function, but the algorithm i'm trying to mimic is 'take a keyword, and remove all instances from the dataset in which that keyword is a substring, until you find the unique set of substrings', which is why this isn't just a straight substitution.

not saying that there's not a more concise way to do this - i think this one of the faster ways? certainly compared to what was happening when this was a huge cross join. but open to nattering about it more

Reply:
I would still say this logic is very C-like instead of Python-like. (i.e. full of indexes into arrays and needing to be careful). What about something like:
```
keywords = sorted(row[1] for row in reader, key=len)
unique_keywords = {}  # casefold -> normal

for keyword in keywords:
  folded = keyword.casefold()
  if not any(x in folded for x in unique_keywords):
    unique_keyword[folded] = keyword
```

Also - when you are checking for a substring - is the logic actually `hasprefix` or `in`?

Reply:
ok, yeah, i think that makes sense.

the operation is an 'in' rather than a 'hasprefix'